R/HDR.R

```diff
@@ -579,3 +580,152 @@ print.hdr <- function(x, ...) {
+  if (length(prob) != 1)                    stop('Error: Probability parameter should be a single numeric value')
+  if (length(approx) != 1)                  stop('Error: Input approx should be a single logical value')
+  if (trials == Inf)                        stop('Error: Input trials must be finite')
+  m <- as.integer(trials)
+  if (m != trials)                          stop('Error: Input trials should be a positive integer')
+  if (m <= 0)                               stop('Error: Input trials should be a positive integer')
+  if (size < Inf) { n <- as.integer(size) } else { n <- Inf }
+  if (n != size)                            stop('Error: Size parameter should be a non-negative integer')
+  if (n < 0)                                stop('Error: Size parameter should be a non-negative integer')
+  if (prob < 0)                             stop('Error: Probability parameter should be between zero and one')
+  if (prob > 1)                             stop('Error: Probability parameter should be between zero and one')
+
+  ########################################################################################
+  ################################# TRIVIAL CASES ########################################
+  ########################################################################################
+
+  #Deal with trivial case where n = 0
+  #Distribution is a point-mass on zero
+  if (n == 0) { 
+    LOGPROBS.TOTAL <- 0 }
+
+  #Deal with trivial case where n = 1
+  #Distribution is a point-mass on m
+  if (n == 1) { 
+    LOGPROBS.TOTAL <- rep(-Inf, m+1)
+    LOGPROBS.TOTAL[m+1] <- 0 }
+
+  #Deal with special case where prob = 1
+  #Distribution is a point-mass on n*m
+  if (prob == 1) {
+    LOGPROBS.TOTAL <- rep(-Inf, n*m+1)
+    LOGPROBS.TOTAL[n*m+1] <- 0 }
+  
+  #Deal with special case where n = Inf and prob > 0
+  #Distribution is a point-mass on infinity
+  if ((n == Inf)&(prob > 0)) {
+    stop('Error: Distribution is a point-mass on infinity') }
+
+  ########################################################################################
+  ############################### NON-TRIVIAL CASES ######################################
+  ########################################################################################
+
+  #Deal with non-trivial cases
+  if ((n > 1)&(prob < 1)) {
+    
+    #Compute distribution of sample total
+    if (approx) {
+      
+      #Compute normal approximation to generalised matching distribution
+      MEAN <- 1 + n*prob - prob^n
+      VAR  <- 1 - prob^(2*n) + n*(prob - prob^2 - prob^(n-1) - prob^n + 2*prob^(n+1))
+      LOGPROBS.TOTAL <- dnorm(0:(n*m), mean = m*MEAN, sd = sqrt(m*VAR), log = TRUE)
+      LOGPROBS.TOTAL <- LOGPROBS.TOTAL - matrixStats::logSumExp(LOGPROBS.TOTAL)
+      
+    } else {
+      
+      #Deal with case where n = Inf and prob = 0
+      if ((n == Inf)&(prob == 0)) {
+        LOGPROBS <- dpois(x, lambda = 1, log = TRUE) }
```

Comment:
@ben-oneill 
R CMD Check says that `x` is not in scope here. This was probably copy-pasted from somewhere else, and you missed an edit. Same issue in `qmatching()`

For HDR, if its asymptotically poission, could just return the result of HDR.poission. Same strategy for qmatching.